topology/src/main/java/io/camunda/zeebe/topology/PersistedClusterTopology.java

```diff
@@ -33,20 +73,89 @@ void update(final ClusterTopology clusterTopology) throws IOException {
     if (this.clusterTopology.equals(clusterTopology)) {
       return;
     }
+    writeToFile(clusterTopology);
+    this.clusterTopology = clusterTopology;
+  }
+
+  public boolean isUninitialized() {
+    return clusterTopology.isUninitialized();
+  }
+
+  private static ClusterTopology readFromFile(
+      final Path topologyFile, final ClusterTopologySerializer serializer) throws IOException {
+    if (!Files.exists(topologyFile)) {
+      return ClusterTopology.uninitialized();
+    }
+
+    final var content = Files.readAllBytes(topologyFile);
+    if (content.length < HEADER_LENGTH) {
+      throw new MissingHeader(topologyFile, content.length);
+    }
+    final var header = ByteBuffer.wrap(content, 0, HEADER_LENGTH).order(ByteOrder.LITTLE_ENDIAN);
+    final var version = header.get();
+    final var expectedChecksum = header.getLong();
+
+    if (version != VERSION) {
+      throw new UnexpectedVersion(topologyFile, version);
+    }
+
+    final var body = Arrays.copyOfRange(content, HEADER_LENGTH, content.length);
 
-    final var serializedTopology = serializer.encode(clusterTopology);
+    final var actualChecksum = checksum(body);
+    if (expectedChecksum != actualChecksum) {
+      throw new ChecksumMismatch(topologyFile, expectedChecksum, actualChecksum);
+    }
+
+    // deserialize the topology
+    return serializer.decodeClusterTopology(body);
+  }
+
+  private void writeToFile(final ClusterTopology clusterTopology) throws IOException {
+    final var body = serializer.encode(clusterTopology);
+    final var checksum = checksum(body);
+    final var buffer =
+        ByteBuffer.allocate(HEADER_LENGTH + body.length)
+            .order(ByteOrder.LITTLE_ENDIAN)
+            .put(VERSION)
+            .putLong(checksum)
+            .put(body);
     Files.write(
         topologyFile,
-        serializedTopology,
+        buffer.array(),
         StandardOpenOption.CREATE,
         StandardOpenOption.WRITE,
         StandardOpenOption.TRUNCATE_EXISTING,
         StandardOpenOption.DSYNC);
+  }
 
-    this.clusterTopology = clusterTopology;
+  private static long checksum(final byte[] bytes) {
+    final var checksum = new CRC32C();
+    checksum.update(bytes);
+    return checksum.getValue();
   }
 
-  public boolean isUninitialized() {
-    return clusterTopology.isUninitialized();
+  private static final class UnexpectedVersion extends RuntimeException {
```

Comment:
:question: If the exception classes are private and only extend `RuntimeException`, is there any use to having them versus just creating `RuntimeException`?